File: webapp/src/main/java/com/box/l10n/mojito/rest/security/UserPreferences.java
package com.box.l10n.mojito.rest.security;

import com.fasterxml.jackson.annotation.JsonInclude;
import java.util.List;

@JsonInclude(JsonInclude.Include.ALWAYS)
public record UserPreferences(
    boolean initialized,
    Integer worksetSize,
    List<String> preferredLocales,
    String shortcutHelp,
    boolean visibleTextEditorEnabled,
    boolean reviewProjectSearchEnabled,
    List<Long> defaultReviewTeamIds,
    String aiReviewProfile,
    boolean aiReviewAutomaticDisabled,
    String aiReviewReasoningEffort,
    String aiReviewPreset,
    String aiReviewStyle,
    Boolean aiReviewShowScore) {

  public UserPreferences {
    aiReviewStyle = aiReviewStyle == null ? "corrections_and_alternatives" : aiReviewStyle;
    aiReviewShowScore = aiReviewShowScore == null ? true : aiReviewShowScore;
    aiReviewProfile = aiReviewProfile == null ? "version_b" : aiReviewProfile;
    aiReviewReasoningEffort = aiReviewReasoningEffort == null ? "low" : aiReviewReasoningEffort;
    if (aiReviewPreset == null) {
      aiReviewPreset =
          "version_a".equals(aiReviewProfile)
              ? "fast"
              : switch (aiReviewReasoningEffort) {
                case "medium" -> "thorough";
                case "high" -> "deep";
                default -> "balanced";
              };
    }
  }

  public UserPreferences(
      boolean initialized,
      Integer worksetSize,
      List<String> preferredLocales,
      String shortcutHelp,
      boolean visibleTextEditorEnabled,
      boolean reviewProjectSearchEnabled,
      List<Long> defaultReviewTeamIds,
      String aiReviewProfile,
      boolean aiReviewAutomaticDisabled,
      String aiReviewReasoningEffort,
      String aiReviewPreset) {
    this(
        initialized,
        worksetSize,
        preferredLocales,
        shortcutHelp,
        visibleTextEditorEnabled,
        reviewProjectSearchEnabled,
        defaultReviewTeamIds,
        aiReviewProfile,
        aiReviewAutomaticDisabled,
        aiReviewReasoningEffort,
        aiReviewPreset,
        null,
        null);
  }

  public UserPreferences(
      boolean initialized,
      Integer worksetSize,
      List<String> preferredLocales,
      String shortcutHelp,
      boolean visibleTextEditorEnabled,
      boolean reviewProjectSearchEnabled,
      List<Long> defaultReviewTeamIds,
      String aiReviewProfile,
      boolean aiReviewAutomaticDisabled,
      String aiReviewReasoningEffort) {
    this(
        initialized,
        worksetSize,
        preferredLocales,
        shortcutHelp,
        visibleTextEditorEnabled,
        reviewProjectSearchEnabled,
        defaultReviewTeamIds,
        aiReviewProfile,
        aiReviewAutomaticDisabled,
        aiReviewReasoningEffort,
        null);
  }

  public UserPreferences(
      boolean initialized,
      Integer worksetSize,
      List<String> preferredLocales,
      String shortcutHelp,
      boolean visibleTextEditorEnabled,
      boolean reviewProjectSearchEnabled,
      List<Long> defaultReviewTeamIds,
      String aiReviewProfile,
      boolean aiReviewAutomaticDisabled) {
    this(
        initialized,
        worksetSize,
        preferredLocales,
        shortcutHelp,
        visibleTextEditorEnabled,
        reviewProjectSearchEnabled,
        defaultReviewTeamIds,
        aiReviewProfile,
        aiReviewAutomaticDisabled,
        "low");
  }

  public static UserPreferences defaults() {
    return new UserPreferences(
        false, null, List.of(), null, false, false, List.of(), "version_b", false, "low");
  }
}

File: webapp/frontend/src/api/userPreferences.ts
export type AiReviewProfile = 'version_a' | 'version_b';
export type AiReviewReasoningEffort = 'low' | 'medium' | 'high';
export type AiReviewPreset = 'fastest' | 'fast' | 'balanced' | 'thorough' | 'deep' | 'ultra';

export type ApiUserPreferences = {
  initialized: boolean;
  worksetSize: number | null;
  preferredLocales: string[];
  shortcutHelp: 'header' | 'bottom' | 'hidden' | null;
  visibleTextEditorEnabled: boolean;
  reviewProjectSearchEnabled: boolean;
  defaultReviewTeamIds: number[];
  aiReviewProfile: AiReviewProfile;
  aiReviewReasoningEffort?: AiReviewReasoningEffort;
  aiReviewPreset?: AiReviewPreset;
  aiReviewAutomaticDisabled: boolean;
};

export type UserPreferencesPatch = Partial<Omit<ApiUserPreferences, 'initialized'>>;

export async function fetchUserPreferences(signal?: AbortSignal): Promise<ApiUserPreferences> {
  const response = await fetch('/api/users/me/preferences', {
    credentials: 'same-origin',
    headers: { Accept: 'application/json' },
    signal,
  });
  if (!response.ok) {
    throw new Error('Could not load your settings.');
  }
  return (await response.json()) as ApiUserPreferences;
}

export async function saveUserPreferences(
  patch: UserPreferencesPatch,
): Promise<ApiUserPreferences> {
  const response = await fetch('/api/users/me/preferences', {
    method: 'PATCH',
    credentials: 'same-origin',
    headers: { Accept: 'application/json', 'Content-Type': 'application/json' },
    body: JSON.stringify(patch),
  });
  if (!response.ok) {
    throw new Error('Could not save your settings. Please try again.');
  }
  return (await response.json()) as ApiUserPreferences;
}
